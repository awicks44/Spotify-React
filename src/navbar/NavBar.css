.menu {
  background-color: #BC4639;
  height: 47px;
  box-shadow: 0 13px 6px -6px #5C2018;
}

.menu .navbarnav {
  color: white;
  font-size: 20px;
  margin: 0px auto;
  padding: 0px;
  width: 400px;
  justify-content: space-between;
}

.stickywrapper {
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1;
}

.categorynav li:hover,
.categorynav li:focus-within {
  background: #BC4639;
  cursor: pointer;
}

.nottransparent {
  background-color: #F3E0DC;
}

.categorynav {
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.3);
  display: none;
  position: absolute;
  z-index: 100;
}

.menu ul {
  list-style: none;
  padding: 16px;
  margin: 0;
}

.menu ul li {
  opacity: 1;
  color: white;
  cursor: pointer;
  transition: 200ms;
  text-decoration: none;
  font-weight: lighter;
  white-space: nowrap;
  font-weight: 500;
}

.menu ul li a {
  opacity: 1;
  color: white;
  cursor: pointer;
  transition: 200ms;
  text-decoration: none;
  font-weight: 600;
  white-space: nowrap;
}

.menu ul li:hover,
.menu ul li a:hover {
  opacity: 1;
}

.menu ul li a,
.menu ul li a a {
  display: flex;
  align-items: center;
  height: 100%;
  width: 100%;
}

.menu ul li {
  padding: auto;
}

.menu>ul li::before {
  content: "";
  width: 0;
  height: 0;
  position: absolute;
  right: 8px;
  top: 50%;
  transform: translateY(-50%);
}

.menu>ul li {
  position: relative;
  margin: 0 8px;
}

.menu>ul li ul {
  visibility: hidden;
  opacity: 0;
  padding: 0;
  min-width: 160px;
  background-color: #D4A59A;
  position: absolute;
  top: calc(var(--menu-height) + 5px);
  left: 50%;
  transform: translateX(-50%);
  transition: 0ms;
  transition-delay: 0ms;
  border: 1.5px solid black;
  border-radius: 4px;
}

.menu>ul li ul li {
  margin: 0;
  padding: 8px 16px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  height: 30px;
  padding-right: 40px;
}

.menu>ul li ul li ul {
  top: -2%;
  left: 100%;
  transform: translate(0);
}

.menu>ul li ul li:hover {
  background-color: #BC4639;
}

.menu>ul li:hover>ul {
  opacity: 1;
  visibility: visible;
  transition-delay: 0ms;
}


.header {
  display: flex;
  justify-content: flex-end;
  margin-bottom: 20px;
  margin-right: 20px;
}

.navbartop {
  position: relative;
  width: 100%;
  height: 150px;
}

.logo {
  text-align: center;
}

.useremail {
  display: block;
  margin-right: -120px;
  color: #4285F4;
  height: 24px;
  right: 50vw;
  background: none;
  border: none;
}

.useremail span:hover {
  background-color: #D4A59A;
  border-radius: 3px;
}

.useremail:focus {
  outline: 0;
}

.emailspan {
  vertical-align: middle;
}

.button {
  background-color: #5C2018 !important;
  color: white !important;
  font-size: 18px !important;
  border: solid !important;
  border-color: #5C2018 !important;
  border-radius: 5px !important;
  display: block !important;
  margin-left: 55px !important;
  left: 10vw !important;
}

.login {
  top: 28%;
  right: 10vw;
  position: absolute;
}

.cart {
  top: 50%;
  right: 20vw;
  position: absolute;
}

.logocontainer {
  position: absolute;
  width: 100%;
  justify-content: center;
}

.loginbtn {
  top: 50%;
  right: 10vw;
  position: absolute;
}

.btnposition {
  margin-right: -30px;
}

.navError {
  margin: auto;
  font-weight: 500;
  color: #FFC107;
}

.cart-image-in {
  position: absolute;
  width: 25%;
  left: 10px;
  bottom: 0px;
  visibility: visible;
}

.cart-image-out {
  position: absolute;
  width: 25%;
  left: 150px;
  top: 25px;
  visibility: visible;
}

.cart-button {
  visibility: hidden;
}

.cartbadge {
  position: absolute;
  left: 91%;
  top: 72px;
  font-weight: bold;
}

.unselectable {
  user-select: none;
}