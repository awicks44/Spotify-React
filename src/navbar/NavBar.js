import React, { useState, useEffect, useCallback } from 'react';
import { Link, useHistory } from 'react-router-dom';
// import UserProfile from '../user profile/UserProfile';
import { useAppContext } from '../utils/Context';
import { QUASAR_API } from '../Constants';
import './NavBar.css';
import Button from '../button/Button';
import cart from './cart.png';
import axios from 'axios';
import { Credentials } from '../Credentials';

// wrapper for the navigation bar, the comments contained in the first section are
// relevent throughout the component
const NavBar = () => {
  const {
    isLoggedIn, setIsLoggedIn, userEmail, setUserEmail, userObject, setUserObject, crt, setCrt
  } = useAppContext();
  const sessionKey = 'user';
  const history = useHistory();
  // const [user, setUser] = useState(sessionStorage.getItem('user'));
  const [demographics, setDemographics] = useState([]);
  const [categories, setCategories] = useState([]);
  const [types, setTypes] = useState([]);
  const [showUserProfile, setShowUserProfile] = useState(false);
  const [error, setError] = useState(false);
  const [token, setToken] = useState('');  
  const spotify = Credentials();  
  const [artists, setArtists] = useState({selectedArtist: '', listOfArtistsFromAPI: []});
  const myArtists = '2aoFQUeHD1U7pL098lRsDU,2pAWfrd7WFF3XhVt9GooDL,77AKJs9SJqxHXbPgtJPKRa';
  const Dante = {url: 'dante_patel', value: '2aoFQUeHD1U7pL098lRsDU', name: 'Dante Patel'};

  useEffect(() => {

    axios('https://accounts.spotify.com/api/token', {
      headers: {
        'Content-Type' : 'application/x-www-form-urlencoded',
        'Authorization' : 'Basic ' + btoa(spotify.ClientId + ':' + spotify.ClientSecret)      
      },
      data: 'grant_type=client_credentials',
      method: 'POST'
    })
    .then(tokenResponse => {      
      setToken(tokenResponse.data.access_token);
        axios(`https://api.spotify.com/v1/artists?ids=${myArtists}`, {
        method: 'GET',
        headers: { 'Authorization' : 'Bearer ' + tokenResponse.data.access_token}
      })
      .then (artistResponse => {        
        console.log(artistResponse);
        setArtists({
          selectedArtist: artists.selectedArtist,
          // listOfArtistsFromAPI: artistResponse.data.categories.items
          listOfArtistsFromAPI: artistResponse.data.artists
        })
      });
      
    });

  }, [artists.selectedArtist, spotify.ClientId, spotify.ClientSecret]); 

  /**
   * Fetches the user's data if logged in and it is needed
   */
  const fetchUserData = async () => {
    if (userObject || !isLoggedIn) {
      return;
    }
    await fetch(`${QUASAR_API}/customers?email=${userEmail}`, {
      method: 'GET',
      headers: new Headers({
        token: sessionStorage.getItem('token')
      })
    }).then(async (response) => {
      if (response.ok) {
        const userData = await response.json();
        setUserObject(userData);
        setCrt(userData.shoppingCart.length);
        sessionStorage.setItem('id', userData.id);
        sessionStorage.setItem('user', userData.email);
      }
    });
  };

  const fetchUserDataCallback = useCallback(fetchUserData,
    [setUserObject, userEmail, userObject, isLoggedIn, setCrt]);
  useEffect(fetchUserDataCallback, [fetchUserDataCallback]);

  useEffect(() => {
    const getDistinctValues = async () => {
      try {
        const demographicsReq = await fetch(`${QUASAR_API}/products/distinctValue/demographics`, {
          headers: new Headers({
            'Content-Type': 'application/json'
          })
        });
        const demographicsData = await demographicsReq.json();
        setDemographics(demographicsData);

        const categoriesReq = await fetch(`${QUASAR_API}/products/distinctValue/categories`, {
          headers: new Headers({
            'Content-Type': 'application/json'
          })
        });
        const categoriesData = await categoriesReq.json();
        setCategories(categoriesData);

        const typesReq = await fetch(`${QUASAR_API}/products/distinctValue/types`, {
          headers: new Headers({
            'Content-Type': 'application/json'
          })
        });
        const typesData = await typesReq.json();
        setTypes(typesData);
      } catch (err) {
        setError(true);
      }
    };
    getDistinctValues();
  }, []);

  /**
   * @name logOut
   * @description logs a user out of the website and returns them to the homepage
   */
  const logOut = () => {
    sessionStorage.removeItem(sessionKey);
    sessionStorage.removeItem('token');
    sessionStorage.removeItem('id');
    setIsLoggedIn(sessionStorage.getItem(sessionKey));
    setUserEmail(sessionStorage.getItem(sessionKey));
    setUserObject();
    history.push('/');
    setCrt('');
  };

  let count;
  if (crt === 0 || isLoggedIn === false) {
    count = (<p> </p>);
  } else {
    count = crt;
  }

  return (
    // <div>
    //   {showUserProfile && (
    //     // <UserProfile
    //     //   user={userObject}
    //     //   setUser={setUserObject}
    //     //   toggleCallback={setShowUserProfile}
    //     // />
    //     <div />
    //   )}
      <div>
        <div>
          <div className="logocontainer">
            {/* <div className="logo">
              <a href="/"><img src={cart} alt="logo" /></a>
            </div> */}
          </div>
          <div className="cartbadge">
            <text className="unselectable">{count}</text>
          </div>
          {/* <div className="login">
            {isLoggedIn
              ? (
                <div>
                  <button type="button" className="useremail mb-2" onClick={() => setShowUserProfile(true)}>
                    <span>
                      <i className="fas fa-cog emailspan" />
                      {' '}
                      {userEmail}
                    </span>
                  </button>
                  <Button id="logout" text="Logout" className="button" onClick={logOut} />
                </div>
              )
              : (
                <div>
                  <div className="useremail" />
                  <Button id="login" text="Login" className="button" onClick={() => history.push('/login')} />
                </div>
              )}
          </div> */}
        </div>
        <nav className="navbar-expand-lg menu">
          {/* {error ? <div className="navError">Nav links could not be rendered</div> : ( */}
            <ul className="nav navbarnav" data-testid="dropdown">
              {/* {Array.isArray(demographics) && demographics.map((demographic) => (
                <li className="nav-item" data-testid={`${demographic.toLowerCase()}`}>
                  <a href={`/products/${demographic.toLowerCase()}`} className="navlink">{demographic}</a>
                  <ul data-testid="categories">
                    {categories.map((category) => (
                      <li>
                        <a href={`/products/${demographic.toLowerCase()}/${category.toLowerCase()}`}>{category}</a>
                        <ul>
                          {types.map((type) => <li><a href={`/products/${demographic.toLowerCase()}/${category.toLowerCase()}/${type.toLowerCase()}`}>{type}</a></li>)}
                        </ul>
                      </li>
                    ))}
                  </ul>
                </li>
              ))} */}
              <li>
                <a>Artists</a>
                <ul>
                  <li><a href={`/artists/${Dante.url}`}>Dante Patel</a></li>
                </ul>
              </li>
              <li>
                <a href={`/contact`}>Contact</a>
              </li>  
            </ul>
        </nav>
      {/* </div> */}
     </div>
  );
};

export default NavBar;
