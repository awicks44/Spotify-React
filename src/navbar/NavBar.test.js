import React from 'react';
import {
  render, screen, cleanup, waitFor
} from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import NavBar from './NavBar';

afterEach(cleanup);

fetchMock.enableMocks();

const categories = [
  'Baseball', 'Basketball', 'Boxing', 'Football',
  'Golf', 'Hockey', 'Running', 'Skateboarding', 'Soccer',
  'Weightlifting'
];

const renderWithRouter = (component) => {
  const history = createMemoryHistory();
  return {
    ...render(
      <Router history={history}>
        {component}
      </Router>
    )
  };
};

it('renders without crashing', () => {
  renderWithRouter(
    <NavBar />
  );
  expect(screen.getByRole('button', { name: /Login/i }));
});

it('header component matches snapshot', () => {
  const { asFragment } = renderWithRouter(
    <NavBar />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('has the correct URL', () => {
  global.window = { location: { pathname: null } };
  !expect(global.window.location.pathname).toEqual('/');
});

it('should load category list from api', async () => {
  fetch.mockResponse({ data: categories });
  const { getAllByTestId } = renderWithRouter(<NavBar />);
  const categoryList = await waitFor(() => getAllByTestId('dropdown'));

  expect(categoryList).toHaveLength(1);
});
