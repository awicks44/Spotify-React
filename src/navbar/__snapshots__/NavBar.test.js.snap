// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="stickywrapper"
  >
    <div
      class="nottransparent"
    >
      <div
        class="navbartop"
      >
        <div
          class="logocontainer"
        >
          <div
            class="logo"
          >
            <a
              href="/"
            >
              <img
                alt="logo"
                src="/sportsapp.png"
              />
            </a>
          </div>
        </div>
        <div
          class="cart"
        >
          <p>
            Shopping Cart Placeholder
          </p>
        </div>
        <div
          class="login"
        >
          <div>
            <div
              class="useremail"
            />
            <div
              class="container text-center"
            >
              <button
                class="button btn"
              >
                Login
              </button>
            </div>
          </div>
        </div>
      </div>
      <nav
        class="navbar navbar-expand-lg menu"
      >
        <ul
          class="nav navbarnav"
          data-testid="dropdown"
        />
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
